java-checks/src/main/java/org/sonar/java/checks/regex/UnusedGroupNamesCheck.java

```diff
@@ -0,0 +1,295 @@
+    if (!groupNumber.isPresent()) {
+      return;
+    }
+    Integer groupNumberValue = groupNumber.get();
+    CapturingGroupTree capturingGroupTree = knownGroups.groupsByNumber.get(groupNumberValue);
+    if (capturingGroupTree == null) {
+      return;
+    }
+    String message = String.format(ISSUE_USE_NAME_INSTEAD_OF_NUMBER, capturingGroupTree.getName().orElse("?"));
+    RegexIssueLocation secondary = toLocation(capturingGroupTree, "Group %d", g -> groupNumberValue);
+    reportIssue(arg0, message, null, Collections.singletonList(secondary));
+  }
+
+  private void checkNoSuchName(KnownGroupsCollector knownGroups, ExpressionTree arg0) {
+    Optional<String> groupName = arg0.asConstant(String.class);
+    if (!groupName.isPresent()) {
+      return;
+    }
+    String groupNameValue = groupName.get();
+    if (!knownGroups.groupsByName.keySet().contains(groupNameValue)) {
+      String message = String.format(ISSUE_NO_GROUP_WITH_SUCH_NAME, groupNameValue);
+      List<RegexIssueLocation> secondaries = knownGroups.groupsByName.values()
+        .stream()
+        .map(group -> toLocation(group, "Named group '%s'", g -> g.getName().get()))
+        .collect(Collectors.toList());
+      reportIssue(arg0, message, null, secondaries);
+    }
+  }
+
+  private static RegexCheck.RegexIssueLocation toLocation(CapturingGroupTree group, String message, Function<CapturingGroupTree, Object> arg) {
+    return new RegexCheck.RegexIssueLocation(group, String.format(message, arg.apply(group)));
+  }
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    getAssignedPrivateVariable(mit)
+      .ifPresent(variableSymbol -> collectGroups(regexForLiterals)
+        .ifPresent(knownGroups -> knownPatternsWithGroups.put(variableSymbol, knownGroups)));
+  }
+
+  private static Optional<Symbol.VariableSymbol> getAssignedPrivateVariable(MethodInvocationTree mit) {
+    Tree parent = mit.parent();
+    if (parent.is(Tree.Kind.VARIABLE)) {
+      Symbol variableSymbol = ((VariableTree) parent).symbol();
+      if (isPrivateVariable(variableSymbol)) {
+        return Optional.of((Symbol.VariableSymbol) variableSymbol);
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<KnownGroupsCollector> collectGroups(RegexParseResult regex) {
+    KnownGroupsCollector visitor = new KnownGroupsCollector(regex);
+    visitor.visit(regex);
+    if (!visitor.groupsByName.isEmpty()) {
+      return Optional.of(visitor);
+    }
+    return Optional.empty();
+  }
+
+  private static class KnownGroupsCollector extends RegexBaseVisitor {
+
+    private final RegexTree target;
+    private Map<String, CapturingGroupTree> groupsByName = new HashMap<>();
+    private Map<Integer, CapturingGroupTree> groupsByNumber = new HashMap<>();
+    private boolean usesBackReferences = false;
+
+    public KnownGroupsCollector(RegexParseResult regex) {
```

Comment:
This seems like a good opportunity to re-introduce the `RegexBaseVisitor.before` method and use that instead of the constructor to set `target`. That way you don't have to pass the regex to both the constructor and the `visit` method.

Reply:
Damn, you are right... I'll re-introduce it then. :)